### DISCUSSION TITLE
Week 1: ML with Python

### DISCUSSION THREAD

**Discussion by BekahHW** [deepgram/community#76] [General help]
## Getting Started

The goal this week is to get through five hours of [freeCodeCamp's ML with python](https://www.freecodecamp.org/learn/machine-learning-with-python/) and check-in here. 

## Learning Resources
- [TensorFlow Titanic Example Tutorial](https://www.tensorflow.org/tutorials/estimator/linear)
- [TensorFlow Iris Example Tutorial](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/estimator/premade.ipynb#scrollTo=jPo5bQwndr9P)
- [Hidden Markov Model documentation](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/HiddenMarkovModel)

## Check-in
1. What notebook are you practicing with? (Some options: Google Colab, Jupyter, Hex, Deepnote)
2. What do you know today that you didn't when you started?
3. What are some key terms or ideas from these lessons?
4. What do you need to be clarified or what questions do you have?
5. What are you excited to learn more about?

**Comment by BekahHW** [comment]
## Check-in Questions
### First 5 lessons
1. Google Colab. This was the easiest for me to get started with and since that's what the videos are using I was able to work through the tutorial.
2. So far, this has been a good review. I went through [Gant Laborde's Learning TFJS](https://www.amazon.com/Learning-TensorFlow-js-Powerful-Machine-JavaScript/dp/1492090794) about a year ago, so I had some familiarity. What I'm really learning in these first 5 videos is the differences between python and JS--which I normally work with.
3. Tensor, datatype, rank, degree, linear regression, classification, clustering, Hidden Markov Models, epoch, categorical data.
4. I need to work more in my Colab to get a better understanding of how what I'm hearing translates to what I can do.
5. I'm most familiar with linear regression, so learning more about the other core algorithms is exciting.

### Lessons 6-10
Learning about the different Core Learning Algorithms has been interesting. I need to look for more examples to better understand the use cases. I've used TensorFlow.js for some of the more common use cases like classification and identification and had some fun with one of my kids and the [teachable machine](https://teachablemachine.withgoogle.com/). I might take a look at [TensorFlow Hub](https://www.tensorflow.org/hub) to check out how these are being used.

**Comment by dannyoka** [comment]
### Check-in Questions
___
**First 5 lessons**
1. I'm using VS Code's built in Jupyter extension, though I may find myself switching to Colab at some point.
2. I hadn't really thought too much about the distinction between AI, ML and neural networks, and how AI can be something as simple as a complicated switch statement (at least to my understanding).
3. Tensors, ranks, training and testing data.
4. I am curious to see where different algorithms fit better in different use cases.
5. I've had very little exposure to tensor flow and other ML libraries beyond numpy, pandas and sklearn, so I'm excited to utilize those and utilize machine learning to do more than just create models.

**Reply by BekahHW** [reply]
I didn't realize that VS Code has a built in Jupyter extension! That's a great tip.

**Reply by dsilvadeepal** [reply]
Same here! I might try Jupyter in VS Code first, since I already use VS Code for other stuff.

**Reply by paceaux** [reply]
I got fed up with the VS Code extension. I kept getting "kernel crashed" errors when I tried to import tensorflow.

**Comment by Se0ngjinKim** [comment]
Check-in
---
First 5 lessons,

1. Google Colab.
2. Tensorflow has imbedded plotting libraries like 'pyplotlib'
3. AI Categories(AI, ML, NN), Linear Regression, Tensors
4. I would like to understand many terms like 'Linear Regression' more 
5. Using the tensorflow with real dataset

**Comment by paceaux** [comment]
# Checkin

1. What notebook are you practicing with?  I'm trying with Jupyter and VS Code. I am not enjoying it. 
2. What do you know today that you didn't when you started? A few things:
   * Tensor is an actual object. I thought it was just a cool name
   * Tensors can have many nested shapes/ dimensions
   * There's four starter algorithms for machine learning
3. What are some key terms or ideas from these lessons?
   * You maybe don't have to know the details of the algorithms, but you should probably know what they are and kinda how they work at a high level
   * The four starter algorithms are Linear Regression, Classification, Clustering, and Hidden Markove models
4. What do you need to be clarified or what questions do you have?
   * probably need to figure out how to actually install/run jupyter locally. It's a bit of a nightmare. trying to set up python kernels and install libraries is rapidly making this unfun. 
5. What are you excited to learn more about?
   * Really want to learn how to apply this to some linguistics and language analysis. 

*UPDATE*
I gave up messing with Jupyter and VS Code. VS Code continually gave me kernel errors. if I tried Jupyter directly, I had problems installing the libraries. 

Collab worked within 2 minutes of trying.

**Reply by BekahHW** [reply]
> You maybe don't have to know the details of the algorithms, but you should probably know what they are and kinda how they work at a high level

I think this has been a big takeaway in my learning. I often get caught up in details, but it's ok to be ok with the magic behind the scenes as you're learning.

**Reply by paceaux** [reply]
Yeah, I've got a few books on NLP and they all went on about (Hidden) Markov models and I spent more time than I should've trying to understand them. I'm actively choosing to check out of that for the sake of actually trying to learn something new.

**Reply by paceaux** [reply]
Other things I've realized. 

when using this (whether collab or Jupyter). This code isn't _just_ about using Tensorflow

This is relying on quite a few other libraries to actually help us understand this. 

1. Pandas. This is used to fetch the data and providing some of the data analysis
2. matplotlib. This is the visualization. It's making all of the charts
3. sklearn is Scikit-Learn. It's a machine-learning library.

**Comment by Se0ngjinKim** [comment]
Day 2: the next 5 lessons,
---

1. What notebook are you practicing with? (Some options: Google Colab, Jupyter, Hex, Deepnote)

- Google Colab.

2. What do you know today that you didn't when you started?

- Many terms and functions.  
  Making an input function which is a parameter of a classifier's evaluation.
- Applying the Hidden Markov Models.  
  The weather example was good to understand it.

3. What are some key terms or ideas from these lessons?

- batches, epochs and estimator
- DNN(Deep Neural Network) Classifier
- K-means clustering

4. What do you need to be clarified or what questions do you have?

- Tensorflow functions.  
  There are many functions and codes related to the Tensorflow.
  I just followed the flow of the classifying, training and prediction processes, 
  but I can hardly understand them now.
  It will be needed to take time...

5. What are you excited to learn more about?

- Building a model.  
  I want to apply my example to understand the processes better.

**Comment by dsilvadeepal** [comment]
### Check In: Lessons 1 - 5
What notebook are you practicing with? 
- Colab. I thought of trying Jupyter/VS Code but it seemed so complex to set up..so Colab it is for now.

What do you know today that you didn't when you started?

- The Python/Tensorflow is way more verbose compared to R. 

What are some key terms or ideas from these lessons?

- The different shapes of tensorflow and how they can be transformed. The various libraries from the Data Analysis of Python that are needed to use Tensorflow like numpy, pandas, matplotlib

What do you need to be clarified or what questions do you have?

- I'd need to dive into the common python libraries like pandas & numpy to understand how they work, since data analysis & cleaning are always the first steps that need to be done prior to ML.

What are you excited to learn more about?

- The difference between tensorflow & sklearn and how the model results compare against each other.

**Comment by dominicduffin1** [comment]
## Week 1:

I completed only the first two videos this week, I hope to catch up more next week.

1. I'm using Google Colab. I'd rather do it in VS Code but I decided for now the time investment to set it up was not worth it.
2. I know that in reality neural networks are not modeled on the human brain, a misconception commonly repeated in mainstream media. I now have an understanding of tensors as a data structure and a clearer understanding of vectors, too, which I'd always been slightly fuzzy about in mathematics.
3. That tensors are essentially multidimensional arrays but with constraints for representing data points in graph space.
4. & 5. How tensors are used in practical problems.
